src/core/cgroup.c

```diff
@@ -2761,6 +2773,41 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
                 log_error_errno(r, "Failed to enable cgroup oom event source: %m");
 }
 
+static int unit_check_cgroup_events(Unit *u) {
+        char *values[2] = {};
+        int r;
+
+        assert(u);
+
+        r = cg_get_keyed_attribute(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, "cgroup.events",
+                                   STRV_MAKE("populated", "frozen"), values);
+        if (r < 0)
+                return r;
+
+        /* The cgroup.events notifications can be merged together so act as we saw the given state for the
+         * first time. The functions we call to handle given state are idempotent, which makes them
+         * effectively remember the previous state. */
+        if (values[0]) {
+                if (streq(values[0], "1"))
+                        unit_remove_from_cgroup_empty_queue(u);
+                else
+                        unit_add_to_cgroup_empty_queue(u);
+        }
+
+        /* Disregard freezer state changes due to operations not initiated by us */
```

Comment:
Current behaviour is a an attempt to to implement,

https://github.com/systemd/systemd/pull/13512#issuecomment-573007207

The think the main idea suggested by Lennart is that managers higher up in the hierarchy always (un)freeze entire subtree. Hence the manager running in delegated subtree doesn't need to care about actions initiated from outside. Since after freeze followed by unfreeze its state will be consistent.  

Also if we would act on external events we might end-up unfreezing units that should remain frozen. For example consider `foo.service` with parent `bar.slice`. If we get thaw on bar.slice due to an external "thaw" on its parent slice we would go ahead and unfreeze everything that is under `bar.slice` including `foo.service`. However, we should really only unfreeze `foo.service` if our client (i.e. on from subtree that is managed by us) ask as to do so.

Previous issue is fixable by tracking two state variables, but since Lennart doesn't want us to care about external events anyway I implemented just that which I think also sidesteps the above issue (covered by explicit test case BTW).